README.md
[![Build Status](https://travis-ci.org/bsaul/geex.svg?branch=master)](https://travis-ci.org/bsaul/geex)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/geex)](https://cran.r-project.org/package=geex)
[![Coverage Status](https://img.shields.io/codecov/c/github/bsaul/geex/master.svg)](https://codecov.io/github/bsaul/geex?branch=master)

# Overview

<img src = "https://raw.githubusercontent.com/bsaul/geex/master/inst/logos/geex-icon-pocket_250x250.png" style = "float:right; padding:0 0 10px 5px">

`geex` provides an extensible API for estimating parameters and their covariance from a set of estimating functions (M-estimation). M-estimation theory has a long history (see the [M-estimation bibliography](https://bsaul.github.io/geex/articles/articles/mestimation_bib.html)). For an excellent introduction, see the primer by L.A. Stefanski and D.D. Boos,  "The Calculus of M-estimation" ([The American Statistician (2002), 56(1), 29-38)](http://www.jstor.org/stable/3087324?seq=1#page_scan_tab_contents); [also available here](http://www4.stat.ncsu.edu/~boos/papers/mest6.pdf)). 

M-estimation encompasses a broad swath of statistical estimators and ideas including:

* the empirical "sandwich" variance estimator
* generalized estimating equations (GEE)
* many maximum likelihood estimators
* robust regression
* and many more

`geex` can implement all of these using a user-defined estimating function. 

## Goals

> If you can specify a set of unbiased estimating equations, `geex` does the rest.

The goals of `geex` are simply:

* To minimize the translational distance between a set of estimating functions and R code;
* To return numerically *accurate* point and covariance estimates from a set of unbiased estimating functions.

`geex` does not necessarily aim to be fast nor precise. Such goals are left to the user to implement or confirm.

# Installation

To install the current version:

```
devtools::install_github("bsaul/geex")
```

# Usage

Start with the examples in the [package introduction](https://bsaul.github.io/geex/articles/v00_geex_intro.html) (also accessible in R by `vignette('00_geex_intro')`). 


# Contributing to geex

Please review the [contributing guidelines](https://github.com/bsaul/geex/blob/master/CONTRIBUTING.md). If you have bug reports, feature requests, or other ideas for `geex`, please file an issue or contact [@bsaul](https://github.com/bsaul).

# Citation

If you use `geex` in a project,
please cite the 
[Journal of Statistical Software paper](https://www.jstatsoft.org/article/view/v092i02).

BibTex entry:

```
  @Article{,
    title = {The Calculus of M-Estimation in {R} with {geex}},
    author = {Bradley C. Saul and Michael G. Hudgens},
    journal = {Journal of Statistical Software},
    year = {2020},
    volume = {92},
    number = {2},
    pages = {1--15},
    doi = {10.18637/jss.v092.i02},
  }
```


NEWS.md
# geex 2.0.0

* Removes `lme4` dependency.
* Removes `grab_psiFUN` functions.
These may be moved to separate package in the future.

# geex 1.1.1

* Removes `ivpack` dependency from additional examples vignette
in order to get geex back on CRAN.
* Removes the internal `compute_sum_of_list` and `compute_pairwise_sum_of_list`
functions from exports.

# geex 1.1.0

* adds a `Asolver` argument to `m_estimate` to allow for using a different to compute the inverse of the A ("bread") matrix.

# geex 1.0.11

* renames the `geex-summary` class to `geex_summary`
* adds `ee_i` (observed estimating function values) to the component object
* various other bug fixes

# geex 1.0.5

* Added `grab_design_levels` to find all factor-levels in a design matrix. This output can then be passed to the `xlev` arg in `model.matrix` or more importantly through `grab_psiFun` for the models. This fixes issue 41. 
   - Notably, if this seems to work well, then perhaps `grab_design_levels` could be added to the body of `grab_psiFun` to automatically handle this issue.
* Edited `binomial_integrand` to hopefully handle singleton groups better. It will now attempt to coerce back to a one-row matrix before calling `apply`.

# geex 1.0.4

* fixes issue when using `grab_psiFUN` for `glm` objects and non-grouped data.
* removes option to use weights with `grab_psiFUN`, as it was not clear that was implemented correctly.

# geex 1.0.3

* adds basic examples for most functions, though to be sure, the vignettes provide more useful examples.

# geex 1.0.2

* requires R >= 3.3

# geex 1.0.1

* add a `call` slot to the S4 `geex` object. Now the `update` function can be used to update elements of an `m_estimate` call.

# geex 1.0.0

* implements an `S4` system throughout `geex`
* `estimate_equations` becomes `m_estimate`. See documentation for changes to arguments. Notably, `eeFUN` becomes `estFUN`
* `make_eeFUN` functions become `grab_psiFUN`
* speeds up the summation of list of matrices with `compute_sum_of_list`
* plus many more updates and vignettes

# geex 0.3.0

* adds a `weights` argument to `estimate_equations` for faster computations with grouped data. See the [weights vignette](https://bsaul.github.io/geex/articles/v04_weights.html) for a demonstration.

# geex 0.2.2

* changes names of list items used in a `corrections_list` in `estimate_equations`. Each item of the `correction_list` must itself be a list at least one item: `correctFUN`. Additional arguments to `correctFUN` may be passed via `correctFUN_control`.

# geex 0.2.1

* adds a vignette explaining how to use a different root finding algorithm via `rootFUN` argument in `estimate_equations`
* fixes issues where different `rootFUN`s would not work:
    * `roots` argument no longer needs to be set if `compute_roots = TRUE`. Instead, starting values are passed via the `rootFUN_control` list. 
    * Removes the `start` argument from `compute_eeroots`; set this option in `rootFUN_control`.

# geex 0.2.0

* overhauls which arguments are passed to `estimate_equations` and how these arguments are parsed internally. See this function's documentation for details
* adds `geexex` dataset for use in examples

# geex 0.1.0

* the inital `geex` release


R/geex.R
#' geex: M-estimation API
#'
#' geex provides an extensible API for estimating parameters and their covariance
#' from a set of estimating functions (M-estimation). M-estimation theory has a
#' long history [see reference in the M-estimation bibliography:
#' \url{https://bsaul.github.io/geex/articles/articles/mestimation_bib.html}.
#' For an excellent introduction, see the primer by L.A. Stefanski and D.D. Boos,
#' "The Calculus of M-estimation" (The American Statistician (2002), 56(1), 29-38)
#' (\url{http://www.jstor.org/stable/3087324}).
#'
#' M-estimation encompasses a broad swath of statistical estimators and ideas including:
#'
#' \itemize{
#' \item the empirical "sandwich" variance estimator
#' \item generalized estimating equations (GEE)
#' \item many maximum likelihood estimators
#' \item robust regression
#' \item and many more}
#'
#' geex can implement all of these using a user-defined estimating function.
#'
#' To learn more about geex, see the package vignettes: \code{browseVignettes(package = 'geex')}.
#'
#' @section Goals:
#' If you can specify a set of unbiased estimating equations, geex does the rest.
#' The goals of geex are simply:
#'
#' \itemize{
#' \item To minimize the translational distance between a set of estimating
#' functions and R code;
#' \item To return numerically accurate point and covariance estimates from
#' a set of unbiased estimating functions.
#' }
#'
#' geex does not, by itself, necessarily aim to be fast nor precise. Such goals
#' are left to the user to implement or confirm.
#'
#' @references Saul, Bradley C., and Michael G. Hudgens. (2020).
#' "The Calculus of M-estimation in R with geex." Journal of Statistical Software 92(2), 1-15.
#' \doi{10.18637/jss.v092.i02}.
#'
"_PACKAGE"
